File: OBCSpack/1BCSpack/1BCSpack/solvers/NM01.m
function out = NM01(A,b,sp,pars)
% This code aims at solving the one-bit compressive sensing with form
%
%      min  sum_{i=1}^n (x_i^2+vareps)^{q/2} + lam * sum_{i=1}^m l_{0/1}(epsilon-b_i<A_i,x>)
%
% vareps > 0, 1 > q > 0, lam > 0, b > 0, epsilon>0
% l_{0/1}(t) = 1 if t > 0 and 0 if t <= 0
% b_i is the i-th entry of b
% A_i is the i-th row of A
% =========================================================================
% Inputs:
%     A   : The sample data in \R^{m-by-n}                       (REQUIRED)
%     b   : The binary observation \in R^m, b_i\in{-1,1}         (REQUIRED)    
%     sp  : Sparsity of the ture signal if known                 (REQUIRED)
%           Set it as [] if unknown                               
%     pars: Parameters are all optional                          (OPTIONAL)
%           pars.x0      --  The initial point         (default zeros(n,1))
%           pars.vareps  --  Parameter in the objective       (default 0.5)
%           pars.q       --  Parameter in the objective       (default 0.5)
%           pars.epsilon --  Parameter in the objective      (default 0.15)
%           pars.lam     --  The penalty parameter              (default 1)
%           pars.tau     --  A useful parameter                 (default 1) 
%           pars.maxit   --  Maximum number of iterations    (default 1000)  
% =========================================================================
% Outputs:
%     Out.x:      The solution 
%     Out.obj:    The objective function value
%     Out.time    CPU time
%     Out.iter:   Number of iterations
% =========================================================================
% Written by Shenglong Zhou on 27/06/2021 based on the algorithm proposed in
%     Shenglong Zhou, Lili Pan, Naihua Xiu, Houduo Qi, SIOPT,  
%     Quadratic convergence of smoothing Newton's method for 0/1 loss optimization,
%     SIAM Journal on Optimization, 31(4): 3184–3211, 2021 
% Send your comments and suggestions to <<< slzhou2021@163.com >>>                                  
% WARNING: Accuracy may not be guaranteed!!!!!  
% =========================================================================
warning off; 
t0  = tic; 
if  nargin < 3  
    disp(' Inputs is not enough !!! \n');
    return;
elseif nargin < 4
    pars = []; 
end

[m,n]  = size(A);
negb   = -b;
if  n  < 1e4
    Ab  = negb.*A;
else
    Ab  = spdiags(negb,0,m,m)*A;    
end

[maxit,lam,tau,mu,epsilon,vareps,cgtol,x,disp] = GetParameters(m,n,pars); 
lam         = max(lam, epsilon^2/2/tau);  
[grad,hess] = def_func(pars);
Fnorm       = @(var)norm(var,'fro')^2;
z           = ones(m,1);    
ACC         = zeros(maxit,1); 
if  norm(x) == 0
    Ax      = epsilon*ones(m,1);
else
    Ax      = Ab*x + epsilon;    
end
Axz         = Ax;   
maxAcc      = 0; 
 
if  disp
    fprintf('\n Start to run the solver: NM01\n')
    fprintf('------------------------------------------\n');
    fprintf('  Iter           HamDist          CPUTime \n')
    fprintf('------------------------------------------\n');
end
for iter      = 1:maxit
     
    [T,empT]  = Ttau(Axz,Ax,tau,lam); 
    g         = grad(x,vareps);
    nT        = nnz(T);  
    
    if  empT
        tmp1 = g;
        raw  = 0; 
        err  = Fnorm(g) + raw  + Fnorm(z);
    else      
        P    = Ab(T,:);
        zT   = z(T);
        tmp1 = g + (zT'*P)'; 
        tmp2 = Ax(T);
        raw  = Fnorm(tmp2);     
        err  = Fnorm(tmp1) + raw +  (Fnorm(z) -Fnorm(zT) ); 
    end
        
    sb        = sign(negb.*(Ax-epsilon)); 
    sb(sb==0) = -1;
    acc       = 1-nnz(sb+negb)/m; 
    x0        = x;
    if iter   > 1 && acc<min(0.5,ACC(iter-1))
       acc    = 1-acc; 
       x0     = -x;
    end
    
    ACC(iter)    = acc; 
    if ACC(iter) > maxAcc
       maxAcc    = ACC(iter);
       maxx      = x0;  
    end
    
    stop1 = (iter>5 && acc<ACC(iter-1) && min(ACC(iter-2:iter-1))==1);    
    stop2 = (iter>5 && acc>0.99995 && raw < 1e-5*sqrt(n) && n>500); 

    if ~stop1 && disp 
        fprintf('  %3d          %8.2f           %.3fsec\n',iter,acc*100,toc(t0));  
    end
    
    if (err<1e-4 && stop2) || stop1 ; break;  end  
     
    if  empT
        u      = - g;
        v      = - z;
    else                
        H      = hess(x,vareps); 
         
        if  nT  < n
            H1  = 1./H; 
            rhs = tmp2 - P*(H1.*tmp1);
            if  nT  < 2e3          
                D   = P*(H1.*P');            
                D(1:nT+1:end) = D(1:nT+1:end) + mu; 
                vT  = D\rhs;  
            else    
                fx  = @(var)(mu*var + P*(H1.*(var'*P)'));
                vT  = my_cg(fx,rhs,cgtol,30,zeros(nT,1)); 
            end        
            v   = -z; 
            v(T)= vT;  
            u   = - H1.*(tmp1 + (vT'*P)');      
        else
            rhs = -mu*tmp1-(tmp2'*P)';
            if  n   < 2e3 
                D   = P'*P;  
                D(1:n+1:end) = D(1:n+1:end) + mu*H';  
                u   = D\rhs;  
            else
                fx = @(var)(mu*(H.*var) + ((P*var)'*P)');  
                u  = my_cg(fx,rhs,cgtol,50,zeros(n,1)); 
            end
            v    = -z; 
            v(T) = (tmp2+P*u)/mu;  
        end
    end 
    
    x   = x + u; 
    z   = z + v;  
    Ax  = Ab*x + epsilon; 
      
    Axz = Ax + tau*z;  
    vareps = max(1e-5,vareps/2);
    if mod(iter,5)==0   
       mu  = max(1e-10,mu/2);                                                                          
    end    
end

if  disp
    fprintf('------------------------------------------\n');
end
if acc < ACC(1); x = maxx; end 
out.lam  = z;
out.time = toc(t0);
out.iter = iter;

if ~isempty(sp) % refinement step
    K       = 6;
    [sx,Ts] = maxk(abs(x),sp+K-1);
    HD      = ones(1,K);
    X       = zeros(n,K); 
    if sx(sp)-sx(sp+1) <= 2e-4 
        tem    = Ts(sp:end);
        for i  = 1:K
            X(:,i)          = zeros(n,1);
            X(Ts(1:sp-1),i) = x(Ts(1:sp-1));
            X(tem(i),i)     = x(tem(i));
            X(:,i)          = X(:,i)/norm(X(:,i));
            HD(i)           = nnz(sign(A*X(:,i))-b)/m; 
        end
        [~,i]   = min(HD); 
        out.sol = X(:,i); 
    else
        out.sol           = zeros(n,1);  
        out.sol(Ts(1:sp)) = x(Ts(1:sp))/norm(x(Ts(1:sp))); 
    end   
else
  x       = SparseApprox(x); 
  out.sol = x / norm(x);   
end
 
clear A b A0 B0 P
end

%--------------------------------------------------------------------------
function [maxit,lam,tau,mu,epsilon,vareps,tolcg,x0,disp] = GetParameters(m,n,pars)
    maxit   = 1e3;
    lam     = 1;
    tau     = 1; 
    mu      = 0.05;   
    epsilon = 0.15; 
    vareps  = 0.5; 
    tolcg   = 1e-10*sqrt(max(m,n));
    x0      = zeros(n,1);
    disp  = 1;
    if isfield(pars,'maxit');   maxit   = pars.maxit;    end 
    if isfield(pars,'lam') ;    lam     = pars.lam;      end 
    if isfield(pars,'tau');     tau     = pars.tau;      end
    if isfield(pars,'mu');      mu      = pars.mu;       end
    if isfield(pars,'epsilon'); epsilon = pars.epsilon;  end
    if isfield(pars,'vareps');  vareps  = pars.vareps;   end
    if isfield(pars,'x0');      x0      = pars.x0;       end 
    if isfield(pars,'disp');    disp   = pars.disp;      end
end

%select the index set T----------------------------------------------------
function [T,empT] = Ttau(Axz,Ax,tau,lam)
    tl   = sqrt(tau*lam/2);
    T    = find(abs(Axz-tl)<tl); 
    empT = isempty(T);  
    if  empT
        zp   = Ax(Ax>=0);  
        T    = [];
        if nnz(zp)>0
            s   = ceil(0.01*nnz(zp));   
            tau = (zp(s))^2/2/lam;     
            tl  = sqrt(tau*lam/2);
            T   = find(abs(Ax-tl)<tl);          
        end
        empT = isempty(T); 
    end
end

% Define functions---------------------------------------------------------
function [grad,hess] = def_func(pars)
    %f(x) = (x^2+e)^(q/2)    
    q    = 0.5;     
    if isfield(pars,'q'); q = pars.q;  end 
    q1   = q/2-1;
    q2   = q/2-2;
    q3   = q-1;
    grad = @(t,e)( t.* ((t.*t+e).^q1) );             
    hess = @(t,e)( ((t.*t+e).^q2).*(q3*t.*t+e) ); 
end

% Conjugate gradient method-------------------------------------------------
function x = my_cg(fx,b,cgtol,cgit,x)
    r = b;
    e = sum(r.*r);
    t = e;
    for i = 1:cgit  
        if e < cgtol*t; break; end
        if  i == 1  
            p = r;
        else
            p = r + (e/e0)*p;
        end
        w  = fx(p); 
        a  = e/sum(p.*w);
        x  = x + a * p;
        r  = r - a * w;
        e0 = e;
        e  = sum(r.*r);
    end
    
end

% get the sparse approximation of x----------------------------------------
function sx = SparseApprox(x)
    n       = length(x);
    xo      = x;
    x       = x.*x;
    T       = find(x>1e-2/n);
    [sx,id] = sort(x(T),'descend'); 
    y       = 0;
    nx      = sum(x(T)); 
    nT      = nnz(T);
    t       = zeros(nT-1,1);
    for i   = 1:nT
        if y > 0.5*nx; break; end
        y    = y + sx(i); 
        if i < nT
        t(i) = sx(i)/sx(i+1);
        end
    end
    
    if  i  < nT
        j  = find(t==max(t)); 
        i  = j(1); 
    else
        i  = nT;
    end
    
    if i  > 1
        i = min(nT,10*i);
    else
        i = nT;
    end
    sx = zeros(n,1);
    sx(T(id(1:i))) = xo(T(id(1:i)));
end

File: OBCSpack/OneBCSpack/OneBCSpack/solvers/NM01.py
from __future__ import annotations

import math
import time
from typing import Dict, Tuple

import numpy as np


def NM01(A, b, sp, pars: Dict | None = None):
    """Python translation of ``NM01.m``."""

    if pars is None:
        pars = {}

    A = np.asarray(A, dtype=float)
    b = np.asarray(b, dtype=float).reshape(-1)
    if sp is None:
        raise ValueError('Inputs is not enough !!!')

    m, n = A.shape
    if b.size != m:
        raise ValueError('Dimension mismatch between A and b')

    negb = -b
    Ab = (negb[:, None]) * A if n < 1e4 else (negb[:, None]) * A

    (maxit, lam, tau, mu, epsilon, vareps, cgtol, x, disp) = _get_parameters(m, n, pars)
    lam = max(lam, epsilon * epsilon / (2 * tau))
    grad, hess = _def_func(pars)
    z = np.ones(m)
    if np.linalg.norm(x) == 0:
        Ax = epsilon * np.ones(m)
    else:
        Ax = Ab @ x + epsilon
    Axz = Ax.copy()
    ACC = np.zeros(maxit)
    maxAcc = 0
    maxx = x.copy()

    t0 = time.perf_counter()
    if disp:
        print('\n Start to run the solver: NM01')
        print('------------------------------------------')
        print('  Iter           HamDist          CPUTime ')
        print('------------------------------------------')

    for it in range(1, maxit + 1):
        T, empT = _Ttau(Axz, Ax, tau, lam)
        g = grad(x, vareps)

        if empT:
            tmp1 = g
            raw = 0.0
            err = _fnorm2(g) + raw + _fnorm2(z)
        else:
            P = Ab[T, :]
            zT = z[T]
            tmp1 = g + P.T @ zT
            tmp2 = Ax[T]
            raw = _fnorm2(tmp2)
            err = _fnorm2(tmp1) + raw + (_fnorm2(z) - _fnorm2(zT))

        sb = np.sign(negb * (Ax - epsilon))
        sb[sb == 0] = -1
        acc = 1 - np.count_nonzero(sb + negb) / m
        x0 = x.copy()
        if it > 1 and acc < min(0.5, ACC[it - 2]):
            acc = 1 - acc
            x0 = -x

        ACC[it - 1] = acc
        if acc > maxAcc:
            maxAcc = acc
            maxx = x0

        stop1 = it > 5 and acc < ACC[it - 2] and np.min(ACC[max(0, it - 3):it - 1]) == 1
        stop2 = it > 5 and acc > 0.99995 and raw < 1e-5 * math.sqrt(n) and n > 500

        if not stop1 and disp:
            print(f"  {it:3d}          {acc * 100:8.2f}           {time.perf_counter()-t0:.3f}sec")
        if (err < 1e-4 and stop2) or stop1:
            break

        if empT:
            u = -g
            v = -z
        else:
            H = hess(x, vareps)
            if T.size < n:
                H1 = 1.0 / H
                rhs = tmp2 - P @ (H1 * tmp1)
                if T.size < 2000:
                    D = P @ (H1[:, None] * P.T)
                    D[np.diag_indices(T.size)] += mu
                    vT = np.linalg.solve(D, rhs)
                else:
                    fx = lambda var: mu * var + P @ (H1 * (P.T @ var))
                    vT = _my_cg(fx, rhs, cgtol, 30, np.zeros(T.size))
                v = -z.copy()
                v[T] = vT
                u = -H1 * (tmp1 + P.T @ vT)
            else:
                rhs = -mu * tmp1 - P.T @ tmp2
                if n < 2000:
                    D = P.T @ P
                    D[np.diag_indices(n)] += mu * H
                    u = np.linalg.solve(D, rhs)
                else:
                    fx = lambda var: mu * (H * var) + P.T @ (P @ var)
                    u = _my_cg(fx, rhs, cgtol, 50, np.zeros(n))
                v = -z.copy()
                v[T] = (tmp2 + P @ u) / mu
        x = x + u
        z = z + v
        Ax = Ab @ x + epsilon
        Axz = Ax + tau * z
        vareps = max(1e-5, vareps / 2)
        if it % 5 == 0:
            mu = max(1e-10, mu / 2)

    if disp:
        print('------------------------------------------')
    if acc < ACC[0]:
        x = maxx

    if sp:
        out_sol = _refine_with_sparsity(A, b, x, sp)
    else:
        out_sol = _sparse_approx(x)
        norm = np.linalg.norm(out_sol)
        if norm > 0:
            out_sol /= norm

    out = {
        'sol': out_sol,
        'lam': z,
        'time': time.perf_counter() - t0,
        'iter': it,
    }
    return out


def _get_parameters(m: int, n: int, pars: Dict) -> Tuple[int, float, float, float, float, float, float, np.ndarray, int]:
    maxit = int(pars.get('maxit', 1_000))
    lam = float(pars.get('lam', 1.0))
    tau = float(pars.get('tau', 1.0))
    mu = float(pars.get('mu', 0.05))
    epsilon = float(pars.get('epsilon', 0.15))
    vareps = float(pars.get('vareps', 0.5))
    tolcg = float(pars.get('tolcg', 1e-10 * math.sqrt(max(m, n))))
    x0 = np.asarray(pars.get('x0', np.zeros(n)), dtype=float)
    disp = int(pars.get('disp', 1))
    return maxit, lam, tau, mu, epsilon, vareps, tolcg, x0, disp


def _def_func(pars: Dict):
    q = float(pars.get('q', 0.5))
    q1 = q / 2 - 1
    q2 = q / 2 - 2
    q3 = q - 1

    def grad(t, e):
        return t * (t * t + e) ** q1

    def hess(t, e):
        return (t * t + e) ** q2 * (q3 * t * t + e)

    return grad, hess


def _Ttau(Axz: np.ndarray, Ax: np.ndarray, tau: float, lam: float):
    tl = math.sqrt(tau * lam / 2)
    mask = np.abs(Axz - tl) < tl
    T = np.flatnonzero(mask)
    empT = T.size == 0
    if empT:
        zp = Ax[Ax >= 0]
        if zp.size:
            s = int(math.ceil(0.01 * zp.size))
            tau = (zp[s - 1]) ** 2 / (2 * lam)
            tl = math.sqrt(tau * lam / 2)
            mask = np.abs(Ax - tl) < tl
            T = np.flatnonzero(mask)
            empT = T.size == 0
    return T, empT


def _refine_with_sparsity(A: np.ndarray, b: np.ndarray, x: np.ndarray, sp: int):
    K = 6
    idx = np.argpartition(-np.abs(x), sp + K - 1)[: sp + K - 1]
    sx = np.abs(x[idx])
    order = np.argsort(-sx)
    Ts = idx[order]
    HD = np.ones(K)
    X = np.zeros((x.size, K))
    if sx[order][sp - 1] - sx[order][sp] <= 2e-4:
        subset = Ts[sp - 1: sp - 1 + K]
        for i in range(min(K, subset.size)):
            Xi = np.zeros_like(x)
            Xi[Ts[: sp - 1]] = x[Ts[: sp - 1]]
            Xi[subset[i]] = x[subset[i]]
            norm = np.linalg.norm(Xi)
            if norm > 0:
                Xi /= norm
            X[:, i] = Xi
            HD[i] = np.count_nonzero(np.sign(A @ Xi) - b) / A.shape[0]
        j = np.argmin(HD)
        return X[:, j]
    sol = np.zeros_like(x)
    sol[Ts[:sp]] = x[Ts[:sp]]
    norm = np.linalg.norm(sol)
    if norm > 0:
        sol /= norm
    return sol


def _sparse_approx(x: np.ndarray):
    n = x.size
    xo = x.copy()
    sq = x * x
    T = np.flatnonzero(sq > 1e-2 / n)
    if T.size == 0:
        return xo
    sx = sq[T]
    order = np.argsort(-sx)
    sx_sorted = sx[order]
    y = 0.0
    total = float(np.sum(sx_sorted))
    limit = 0
    for i, val in enumerate(sx_sorted, start=1):
        y += val
        if y > 0.5 * total:
            limit = i
            break
    if limit == 0:
        limit = sx_sorted.size
    ratios = np.zeros(max(0, limit - 1))
    for i in range(limit - 1):
        if sx_sorted[i + 1] != 0:
            ratios[i] = sx_sorted[i] / sx_sorted[i + 1]
    if ratios.size:
        j = np.argmax(ratios) + 1
    else:
        j = limit
    if j > 1:
        j = min(T.size, 10 * j)
    indices = T[order[:j]]
    sparse = np.zeros_like(x)
    sparse[indices] = xo[indices]
    return sparse


def _fnorm2(var: np.ndarray) -> float:
    return float(np.linalg.norm(var) ** 2)


def _my_cg(fx, b: np.ndarray, cgtol: float, cgit: int, x0: np.ndarray):
    x = x0.copy()
    r = b - fx(x)
    e = float(r @ r)
    t = e
    p = r.copy()
    for _ in range(cgit):
        if e < cgtol * t:
            break
        w = fx(p)
        a = e / float(p @ w)
        x = x + a * p
        r = r - a * w
        e0 = e
        e = float(r @ r)
        p = r + (e / e0) * p
    return x
